File: utils/heap.rb

class MaxHeap

  attr_reader :heap

  def initialize(arr = nil)
    @heap = Array(arr)
  end

  def add(value)
    @heap << value
    new_value_index = @heap.size - 1
    parent_index = get_parent_index(new_value_index)

    while parent_index && @heap[parent_index] < @heap[new_value_index]
      swap(parent_index, new_value_index)
      new_value_index = parent_index

      parent_index = get_parent_index(new_value_index)
    end
  end

  def remove(index)
    last_index = @heap.size - 1

    if index == last_index
      @heap.delete_at(last_index)
    else
      swap(index, last_index)
      @heap.delete_at(last_index)

      curr_index = index
      left_index = get_left_child_index(curr_index)
      right_index = get_right_child_index(curr_index)
      while !leaf_node?(curr_index) && !in_correct_position?(curr_index)
        swap_index = index_to_swap(left_index, right_index)
        swap(curr_index, swap_index)
        curr_index = swap_index

        left_index = get_left_child_index(curr_index)
        right_index = get_right_child_index(curr_index)
      end
    end
  end

  def empty?
    @heap.empty?
  end

  private

  def index_to_swap(left_index, right_index)
    return left_index if @heap[right_index].nil?
    return right_index if @heap[left_index].nil?

    @heap[left_index] >= @heap[right_index] ? left_index : right_index
  end

  def in_correct_position?(index)
    in_correct_left_position?(index) && in_correct_right_position?(index)
  end

  def in_correct_left_position?(index)
    left_index = get_left_child_index(index)
    @heap[left_index].nil? || @heap[index] >= @heap[left_index]
  end

  def in_correct_right_position?(index)
    right_index = get_right_child_index(index)
    @heap[right_index].nil? || @heap[index] >= @heap[right_index]
  end

  def leaf_node?(index)
    @heap[get_left_child_index(index)].nil? && @heap[get_right_child_index(index)].nil?
  end

  def swap(index, index2)
    tmp = @heap[index]
    @heap[index] = @heap[index2]
    @heap[index2] = tmp
  end

  def get_left_child_index(index)
    2 * index + 1
  end

  def get_right_child_index(index)
    2 * (index + 1)
    # 2 * index + 2
  end

  def get_parent_index(index)
    return nil if index == 0

    ((index - 1) / 2.0).floor
  end

end


hi = MaxHeap.new([10, 7, 6])
hi.add(8)
hi.add(3)
hi.add(12)
hi.add(11)
hi.add(5)
puts hi.inspect

while !hi.empty?
  puts hi.heap[0]
  hi.remove(0)
end

File: js_output/utils/heap.js
import assert from "./assert.js";
class Heap {
    isMax;
    array;
    constructor(isMax) {
        this.isMax = isMax;
        this.array = [];
    }
    add(num) {
        if (this.array.length === 0) {
            this.array.push(num);
            return;
        }
        this.array.push(num);
        let currIndex = this.array.length - 1;
        while (true) {
            if (currIndex === 0)
                break;
            const parentIndex = Math.floor((currIndex - 1) / 2);
            if (this.isMax
                ? this.array[currIndex] > this.array[parentIndex]
                : this.array[currIndex] < this.array[parentIndex]) {
                this.swap(currIndex, parentIndex);
                currIndex = parentIndex;
            }
            else {
                break;
            }
        }
    }
    remove(index) {
        if (this.array[index] === undefined)
            throw "Element not found";
        this.swap(index, this.array.length - 1);
        this.array.pop();
        let currIndex = index;
        while (true) {
            const leftIndex = 2 * currIndex + 1;
            const rightIndex = 2 * currIndex + 2;
            const indexToSwap = this.getIndexToSwap(this.array[currIndex], leftIndex, rightIndex);
            if (indexToSwap !== -1) {
                this.swap(currIndex, indexToSwap);
                currIndex = indexToSwap;
            }
            else {
                break;
            }
        }
    }
    getIndexToSwap(currValue, leftIndex, rightIndex) {
        let indexToSwap = -1;
        let minOrMax = currValue;
        const leftValue = this.array[leftIndex];
        const rightValue = this.array[rightIndex];
        if (leftValue !== undefined) {
            if (this.isMax ? leftValue > minOrMax : leftValue < minOrMax) {
                indexToSwap = leftIndex;
                minOrMax = leftValue;
            }
        }
        if (this.array[rightIndex] !== undefined) {
            if (this.isMax ? rightValue > minOrMax : rightValue < minOrMax) {
                indexToSwap = rightIndex;
                minOrMax = rightValue;
            }
        }
        return indexToSwap;
    }
    swap(index1, index2) {
        const tmp = this.array[index1];
        this.array[index1] = this.array[index2];
        this.array[index2] = tmp;
    }
    toString() {
        return this.array.join(",");
    }
}
const heap = new Heap(true);
heap.add(5);
heap.add(4);
heap.add(3);
heap.add(2);
heap.add(1);
heap.add(2);
heap.add(8);
heap.add(5);
assert(heap.toString() === "8,5,5,4,1,2,3,2");
heap.remove(1);
assert(heap.toString() === "8,4,5,2,1,2,3");
heap.remove(1);
assert(heap.toString() === "8,3,5,2,1,2");
const heap2 = new Heap(false);
heap2.add(5);
heap2.add(4);
heap2.add(3);
heap2.add(2);
heap2.add(1);
heap2.add(2);
heap2.add(8);
heap2.add(5);
assert(heap2.toString() === "1,2,2,5,3,4,8,5");
heap2.remove(2);
assert(heap2.toString() === "1,2,4,5,3,5,8");
heap2.remove(5);
assert(heap2.toString() === "1,2,4,5,3,8");
console.log("All assertions passed.");
